### PR TITLE
pim6: Adding new show cli

### PR DESCRIPTION
**Pull request by SaiGomathiN**
Adding 
show ipv6 pim channel
show ipv6 pim group-type
show ipv6 pim interface
show ipv6 pim join
show ipv6 pim jp-agg
show ipv6 pim local-membership
show ipv6 pim neighbor
show ipv6 pim nexthop
show ipv6 pim nexthop-lookup
commands

### CODE REVIEW

File: pimd/pim6_cmd.c [hunk 1]
```diff
@@ -43,7 +43,32 @@
...
+       "show ipv6 pim [vrf NAME] channel [json]",
+       SHOW_STR
+       IPV6_STR
+       PIM_STR
+       VRF_CMD_HELP_STR
+       "PIM downstream channel info\n"
+       JSON_STR)
+{
+	struct vrf *vrf_struct;
+	bool uj = use_json(argc, argv);
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c]: This can be removed now. Use json argument directly instead.

File: pimd/pim_cmd_common.c [hunk 2]
```diff
@@ -507,8 +508,8 @@ static void pim_show_channel_helper(struct pim_instance *pim, struct vty *vty,
 		json_object_object_add(json_group, src_str, json_row);
 
 	} else {
-		vty_out(vty, "%-16s %-15s %-15s %-10s %-5s %-10s %-11s %-6s\n",
-			ch->interface->name, src_str, grp_str,
+		vty_out(vty, "%-16s %-15pPAs %-15pPAs %-10s %-5s %-10s %-11s %-6s\n",
```
Comment (mobash-rasool) [inline pimd/pim_cmd_common.c]: Good. I am expecting this kind of changes all over.

File: pimd/pim_cmd_common.c [hunk 3]
```diff
@@ -469,14 +469,15 @@ static void pim_show_channel_helper(struct pim_instance *pim, struct vty *vty,
...
 	json_object *json_row = NULL;
 
-	pim_inet4_dump("<src?>", up->sg.src, src_str, sizeof(src_str));
-	pim_inet4_dump("<grp?>", up->sg.grp, grp_str, sizeof(grp_str));
-
 	if (uj) {
+		char grp_str[PIM_ADDRSTRLEN];
+		char src_str[PIM_ADDRSTRLEN];
+
+		snprintfrr(grp_str, sizeof(grp_str), "%pPAs", &up->sg.grp);
```
Comment (mobash-rasool) [inline pimd/pim_cmd_common.c:459]: Can you check once If we need this here, explore json_object_string_addf api once.

File: pimd/pim_cmd_common.c [hunk 4]
```diff
@@ -1079,6 +1080,9 @@ void ip_pim_ssm_show_group_range(struct pim_instance *pim, struct vty *vty,
 		vty_json(vty, json);
 	} else
 		vty_out(vty, "SSM group range : %s\n", range_str);
+#else
+	/* TBD with PIM6_SSM_STANDARD_RANGE */
```
Comment (mobash-rasool) [inline pimd/pim_cmd_common.c]: You can define FF3x::/96 as PIM_SSM_STANDARD_RANGE for PIMv6.
#if PIM_IPV == 4
#define PIM_SSM_STANDARD_RANGE "232.0.0.0/8"
#else
#define PIM_SSM_STANDARD_RANGE "FF3x::/96"

**Review by mobash-rasool — CHANGES_REQUESTED** [review]

File: pimd/pim_rp.c [hunk 5]
```diff
@@ -1135,30 +1135,32 @@ struct pim_rpf *pim_rp_g(struct pim_instance *pim, struct in_addr group)
...
 				   __func__);
 		up->s_addr = INADDR_ANY;
 		return 0;
 	}
 
-	*up = (source.s_addr == INADDR_ANY) ? rp_info->rp.rpf_addr.u.prefix4
+	*up = (pim_addr_is_any(source)) ? rp_info->rp.rpf_addr.u.prefix4
 					    : source;
+#else
+	/* TBD */
```
Comment (mobash-rasool) [inline pimd/pim_rp.c]: I have done this change in my static RP PR.

File: pimd/pim6_cmd.c [hunk 6]
```diff
@@ -388,6 +389,95 @@ DEFPY (show_ipv6_pim_neighbor_vrf_all,
...
+	char nexthop_addr_str[PIM_ADDRSTRLEN];
+	char grp_str[PIM_ADDRSTRLEN];
+	struct vrf *vrf_struct;
+
+	vrf_struct = vrf_lookup_by_name(vrf ? vrf : VRF_DEFAULT_NAME);
+
+	if (!vrf_struct)
+		return CMD_WARNING;
+
+	result = inet_pton(AF_INET6, addr_str, &src_addr);
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c]: This may not be required here since we are already getting the src_addr as addr itself, you can try running the command without this check and see if the CLI is validating this check already.

File: pimd/pim6_cmd.c [hunk 7]
```diff
@@ -388,6 +389,95 @@ DEFPY (show_ipv6_pim_neighbor_vrf_all,
...
+	if (result <= 0) {
+		vty_out(vty, "Bad unicast address %s: errno=%d: %s\n", addr_str,
+			errno, safe_strerror(errno));
+		return CMD_WARNING;
+	}
+
+	result = inet_pton(AF_INET6, addr1_str, &grp_addr);
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c]: This may not be required since clip does it for you. You have it already in addr1.

File: pimd/pim_cmd_common.c (hunk 4 above)
Comment (SaiGomathiN) [reply pimd/pim_cmd_common.c]: Fixed, Thankyou

File: pimd/pim_rp.c (hunk 5 above)
Comment (SaiGomathiN) [reply pimd/pim_rp.c]: Done

File: pimd/pim6_cmd.c [hunk 8]
```diff
@@ -33,16 +33,445 @@
...
+		return CMD_WARNING;
+
+	ip_pim_ssm_show_group_range(vrf_struct->info, vty, uj);
+
+	return CMD_SUCCESS;
+}
+
+DEFPY (show_ipv6_pim_interface,
+       show_ipv6_pim_interface_cmd,
+       "show ipv6 pim [mlag$mlag] [vrf NAME] interface [[detail|WORD]$detail] [json$json]",
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c]: Do we need mlag here? Lets remove just the optional mlag keyword here and pass false to the api  pim_show_interfaces_single

File: pimd/pim6_cmd.c [hunk 9]
```diff
@@ -33,16 +33,445 @@
...
+
+	if (uj)
+		vty_json(vty, json_parent);
+
+	return CMD_SUCCESS;
+}
+
+DEFPY (show_ipv6_pim_interface_vrf_all,
+       show_ipv6_pim_interface_vrf_all_cmd,
+       "show ipv6 pim [mlag$mlag] vrf all interface [[detail|WORD]$detail] [json$json]",
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c]: Remove mlag

File: pimd/pim6_cmd.c [hunk 10]
```diff
@@ -33,16 +33,445 @@
...
+       JSON_STR)
+{
+	pim_sgaddr sg = {0};
+	struct vrf *vrf_struct;
+	json_object *json_parent = NULL;
+	json_object *json_vrf = NULL;
+
+	if (json)
+		json_parent = json_object_new_object();
+	RB_FOREACH (vrf_struct, vrf_name_head, &vrfs_by_name) {
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c:239]: Add an extra line above this statement, it looks odd now.

File: pimd/pim6_cmd.c [hunk 11]
```diff
@@ -33,16 +33,445 @@
...
+
+	if (json)
+		vty_json(vty, json_parent);
+
+	return CMD_SUCCESS;
+}
+
+DEFPY (show_ipv6_pim_neighbor_vrf_all,
+       show_ipv6_pim_neighbor_vrf_all_cmd,
+       "show ipv6 pim vrf all neighbor [[detail|WORD]$detail] [json$json]",
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c:345]: can you check $detail keyword here, you need it only for detail or both for word and detail?

File: pimd/pim6_cmd.c [hunk 12]
```diff
@@ -33,16 +33,445 @@
...
+	struct pim_nexthop nexthop;
+	char nexthop_addr_str[PIM_ADDRSTRLEN];
+	struct vrf *vrf_struct;
+
+	vrf_struct = vrf_lookup_by_name(vrf ? vrf : VRF_DEFAULT_NAME);
+
+	if (!vrf_struct)
+		return CMD_WARNING;
+
+	memcpy(&src_addr, &source, sizeof(source));
```
Comment (mobash-rasool) [inline pimd/pim6_cmd.c]: you don't need these src_addr and grp_addr separately. You can use source and group directly.

File: pimd/pim_cmd.c [hunk 13]
```diff
@@ -1613,7 +906,8 @@ static void pim_show_interface_traffic_single(struct pim_instance *pim,
 			json_object_object_add(json, ifp->name, json_row);
 		} else {
 			vty_out(vty,
-				"%-16s %8u/%-8u %7u/%-7u %7u/%-7u %7u/%-7u %7u/%-7u %7u/%-7u %7" PRIu64 "/%-7" PRIu64 "\n",
+				"%-16s %8u/%-8u %7u/%-7u %7u/%-7u %7u/%-7u %7u/%-7u %7u/%-7u %7" PRIu64
```
Comment (mobash-rasool) [inline pimd/pim_cmd.c]: Why the change is moved to next line here?

File: pimd/pim_cmd.c [hunk 14]
```diff
@@ -3760,7 +2469,6 @@ static void clear_interfaces(struct pim_instance *pim)
 		return CMD_WARNING_CONFIG_FAILED;			\
 	}
 
-#if PIM_IPV != 6
```
Comment (mobash-rasool) [inline pimd/pim_cmd.c:3763]: Don't touch the MSDP changes for now, we don't know it needs to be implemented or not, we are not going to implement it so lets just remove the commits related to MSDP, keep it in some branch if required later we can raise it from there.

File: pimd/pim_cmd.c [hunk 15]
```diff
@@ -4478,9 +3185,9 @@ DEFUN (show_ip_pim_assert_winner_metric,
 	return CMD_SUCCESS;
 }
 
-DEFUN (show_ip_pim_interface,
+DEFPY (show_ip_pim_interface,
        show_ip_pim_interface_cmd,
-       "show ip pim [mlag] [vrf NAME] interface [detail|WORD] [json]",
+       "show ip pim [mlag$mlag] [vrf NAME] interface [[detail|WORD]$detail] [json$json]",
```
Comment (mobash-rasool) [inline pimd/pim_cmd.c]: is $detail applicable to both detail and word?

File: pimd/pim_cmd.c [hunk 16]
```diff
@@ -11147,14 +9800,12 @@ void pim_cmd_init(void)
 	install_element(ENABLE_NODE, &no_debug_pim_mlag_cmd);
 	install_element(ENABLE_NODE, &debug_pim_vxlan_cmd);
 	install_element(ENABLE_NODE, &no_debug_pim_vxlan_cmd);
-#if PIM_IPV != 6
```
Comment (mobash-rasool) [inline pimd/pim_cmd.c:11150]: why is this removed?

File: pimd/pim_cmd.c [hunk 17]
```diff
@@ -11200,20 +9851,17 @@ void pim_cmd_init(void)
 	install_element(CONFIG_NODE, &no_debug_pim_mlag_cmd);
 	install_element(CONFIG_NODE, &debug_pim_vxlan_cmd);
 	install_element(CONFIG_NODE, &no_debug_pim_vxlan_cmd);
-#if PIM_IPV != 6
```
Comment (mobash-rasool) [inline pimd/pim_cmd.c:11203]: why is this removed?

File: pimd/pim_cmd.c [hunk 18]
```diff
@@ -11200,20 +9851,17 @@ void pim_cmd_init(void)
...
 	install_element(CONFIG_NODE, &no_debug_msdp_events_cmd);
 	install_element(CONFIG_NODE, &debug_msdp_packets_cmd);
 	install_element(CONFIG_NODE, &no_debug_msdp_packets_cmd);
-#endif /* PIM_IPV != 6 */
 	install_element(CONFIG_NODE, &debug_mtrace_cmd);
 	install_element(CONFIG_NODE, &no_debug_mtrace_cmd);
 	install_element(CONFIG_NODE, &debug_bsm_cmd);
 	install_element(CONFIG_NODE, &no_debug_bsm_cmd);
 
-#if PIM_IPV != 6
```
Comment (mobash-rasool) [inline pimd/pim_cmd.c:11216]: why is this removed?

File: pimd/pim_vty.c [hunk 19]
```diff
@@ -448,7 +471,6 @@ int pim_interface_config_write(struct vty *vty)
 #endif
 			}
 			if_vty_config_end(vty);
-
```
Comment (patrasar) [inline pimd/pim_vty.c:451]: Please remove this line

File: pimd/pim_cmd_common.c [hunk 20]
```diff
@@ -343,7 +341,7 @@ static void pim_show_membership_helper(struct vty *vty,
 	}
 
 	json_row = json_object_new_object();
-	json_object_string_add(json_row, "source", ch_src_str);
+	json_object_string_addf(json_row, "source", "%pPAs", &ch->sg.src);
 	json_object_string_add(json_row, "group", ch_grp_str);
```
Comment (mobash-rasool) [inline pimd/pim_cmd_common.c:345]: you can use json_object_string_addf for group as well right?

File: pimd/pim_cmd_common.c [hunk 21]
```diff
@@ -444,8 +440,6 @@ void pim_show_membership(struct pim_instance *pim, struct vty *vty, bool uj)
 			}
 		}
 	}
-
-	json_object_free(json);
```
Comment (mobash-rasool) [inline pimd/pim_cmd_common.c]: There is a memleak here. Consider uj is passed as false. then still json Is malloced at line 362 and is never freed since you have removed it.

File: pimd/pim_vty.c [hunk 22]
```diff
@@ -364,18 +374,29 @@ int pim_config_write(struct vty *vty, int writes, struct interface *ifp,
 	if (pim_ifp->gm_join_list) {
 		struct listnode *node;
 		struct gm_join *ij;
-		for (ALL_LIST_ELEMENTS_RO(pim_ifp->gm_join_list, node, ij)) {
+		for (ALL_LIST_ELEMENTS_RO(
```
Comment (patrasar) [inline pimd/pim_vty.c:377]: It seems you have not done any changes in this functions, it seems to be just style issues

Can we remove these in this function

File: pimd/pim_vty.c [hunk 23]
```diff
@@ -293,27 +293,30 @@ int pim_config_write(struct vty *vty, int writes, struct interface *ifp,
 	}
```
Comment (patrasar) [inline pimd/pim_vty.c:293]: If there is no changes in pim_vty.c except style issues, Please dont include these in this PR.

File: pimd/pim_rp.c [hunk 24]
```diff
@@ -1135,30 +1135,34 @@ struct pim_rpf *pim_rp_g(struct pim_instance *pim, struct in_addr group)
...
 				   __func__);
-		up->s_addr = INADDR_ANY;
+		*up = PIMADDR_ANY;
 		return 0;
 	}
 
-	*up = (source.s_addr == INADDR_ANY) ? rp_info->rp.rpf_addr.u.prefix4
-					    : source;
+	if (pim_addr_is_any(source))
+#if PIM_IPV == 4 || !defined(PIM_V6_TEMP_BREAK)
```
Comment (patrasar) [inline pimd/pim_rp.c:1159]: Please use pim_addr_from_prefix

File: pimd/pim_cmd_common.c [hunk 25]
```diff
@@ -0,0 +1,1356 @@
...
+		!pim_addr_cmp(match.grp, item.grp)) &&
+	       (pim_addr_is_any(match.src) ||
+		!pim_addr_cmp(match.src, item.src));
+}
+struct vrf *pim_cmd_lookup_vrf(struct vty *vty, struct cmd_token *argv[],
+				      const int argc, int *idx)
+{
+	struct vrf *vrf;
+
+	if (argv_find(argv, argc, "NAME", idx))
```
Comment (patrasar) [inline pimd/pim_cmd_common.c:60]: pass the VRF name in the caller , for default it will be NULL.
Since we are using DEFPY now, no need ofpassing argv and argc and idx

File: pimd/pim6_cmd.c [hunk 26]
```diff
@@ -33,16 +33,445 @@
...
+       SHOW_STR
+       IPV6_STR
+       PIM_STR
+       VRF_CMD_HELP_STR
+       "PIM neighbor information\n"
+       "Detailed output\n"
+       "Name of interface or neighbor\n"
+       JSON_STR)
+{
+	struct vrf *vrf_struct;
```
Comment (patrasar) [inline pimd/pim6_cmd.c:321]: Please keep the name as vrf, we use vrf everywhere in other places

File: pimd/pim6_cmd.c [hunk 27]
```diff
@@ -33,16 +33,445 @@
...
+
+	if (!vrf_struct)
+		return CMD_WARNING;
+
+	if (json)
+		json_parent = json_object_new_object();
+	if (detail)
+		pim_show_neighbors_single(vrf_struct->info, vty, detail,
+					  json_parent);
+	else
```
Comment (patrasar) [inline pimd/pim6_cmd.c:334]: should we move all these in a common wrapper as we are doing for cli for IPV4 and IPV6

File: pimd/pim6_cmd.c [hunk 28]
```diff
@@ -33,16 +33,445 @@
...
+	if (!vrf_struct)
+		return CMD_WARNING;
+
+	memcpy(&src_addr, &source, sizeof(source));
+	memcpy(&grp_addr, &group, sizeof(group));
+	if (!IPV6_ADDR_SAME(&source, &in6addr_any))
+		memcpy(&src_addr, &source, sizeof(source));
+	if (!IPV6_ADDR_SAME(&group, &in6addr_any))
+		memcpy(&grp_addr, &group, sizeof(group));
+
```
Comment (patrasar) [inline pimd/pim6_cmd.c:435]: Same as above, Might be we will get comment to have a common wrapper for IPV4 and IPV6

File: pimd/pim_cmd_common.c [hunk 29]
```diff
@@ -0,0 +1,1356 @@
...
+	ifindex_t first_ifindex;
+	struct interface *ifp = NULL;
+	char buf[PREFIX_STRLEN];
+
+	for (nh_node = pnc->nexthop; nh_node; nh_node = nh_node->next) {
+		first_ifindex = nh_node->ifindex;
+		ifp = if_lookup_by_index(first_ifindex, pim->vrf->vrf_id);
+
+		vty_out(vty, "%-15s ",
+			inet_ntop(AF_INET, &pnc->rpf.rpf_addr.u.prefix4, buf,
```
Comment (patrasar) [inline pimd/pim_cmd_common.c:1068]: code is missing for IPV6. please check.
Print PIM_AF

File: pimd/pim_cmd_common.c [hunk 30]
```diff
@@ -0,0 +1,1356 @@
...
+
+void pim_show_nexthop(struct pim_instance *pim, struct vty *vty)
+{
+	struct pnc_cache_walk_data cwd;
+
+	cwd.vty = vty;
+	cwd.pim = pim;
+	vty_out(vty, "Number of registered addresses: %lu\n",
+		pim->rpf_hash->count);
+	vty_out(vty, "Address         Interface        Nexthop\n");
```
Comment (patrasar) [inline pimd/pim_cmd_common.c:1085]: For IPV6 address we might need more space.

Please check every show command where we are displaying address

File: pimd/pim_cmd.c (hunk 14 above)
Comment (SaiGomathiN) [reply pimd/pim_cmd.c:3763]: This was cherry-picked from your changes, i didn't change anything

File: pimd/pim_cmd.c (hunk 16 above)
Comment (SaiGomathiN) [reply pimd/pim_cmd.c:11150]: The whole commit was cherry-picked from your changes, i didn't change anything

File: pimd/pim6_cmd.h [hunk 31]
```diff
@@ -39,14 +39,5 @@
 #define DEBUG_MLD_TRACE_STR "MLD internal daemon activity\n"
 
 void pim_cmd_init(void);
-/*
```
Comment (patrasar) [inline pimd/pim6_cmd.h:42]: The branch is not rebase to master, why this file is present?
Please rebase, might we get some conflicts

File: pimd/pim_cmd.c [hunk 32]
```diff
@@ -4491,69 +3198,71 @@ DEFUN (show_ip_pim_interface,
        "interface name\n"
        JSON_STR)
 {
-	int idx = 2;
-	struct vrf *vrf = pim_cmd_lookup_vrf(vty, argv, argc, &idx);
-	bool uj = use_json(argc, argv);
-	bool mlag = false;
+	struct vrf *vrf_struct;
+	bool uj = !!json;
```
Comment (patrasar) [inline pimd/pim_cmd.c:3201]: What is this symbol "!!"

File: pimd/pim_cmd.c (hunk 17 above)
Comment (mobash-rasool) [reply pimd/pim_cmd.c:11203]: oh ok ok got it, since we have made the files itself separate so its no problem.

**Review by mobash-rasool — COMMENTED** [review]
Do you need the first 3 commits for your changes? If not, lets not add it in the community PR

File: pimd/pim6_cmd.c (hunk 8 above)
Comment (SaiGomathiN) [reply pimd/pim6_cmd.c]: Fixed, Thankyou

File: pimd/pim6_cmd.c (hunk 9 above)
Comment (SaiGomathiN) [reply pimd/pim6_cmd.c]: Done

File: pimd/pim6_cmd.c (hunk 10 above)
Comment (SaiGomathiN) [reply pimd/pim6_cmd.c:239]: Done

File: pimd/pim_cmd.c (hunk 32 above)
Comment (SaiGomathiN) [reply pimd/pim_cmd.c:3201]: As we are assigning const char* to bool, safe to use !!, (followed from Donald's changes)
